File: cryptography-cipher/src/main/java/dev/retrotv/crypto/cipher/block/mode/OFB.java
package dev.retrotv.crypto.cipher.block.mode;

import dev.retrotv.crypto.cipher.block.BlockCipher;
import dev.retrotv.crypto.cipher.block.StreamBlockCipherMode;
import dev.retrotv.crypto.cipher.enums.EMode;
import dev.retrotv.crypto.cipher.param.Param;
import dev.retrotv.crypto.cipher.param.ParamWithIV;
import dev.retrotv.crypto.cipher.result.Result;
import dev.retrotv.crypto.exception.CryptoFailException;
import lombok.NonNull;

import org.bouncycastle.crypto.CipherParameters;
import org.bouncycastle.crypto.modes.OFBBlockCipher;
import org.bouncycastle.crypto.params.KeyParameter;
import org.bouncycastle.crypto.params.ParametersWithIV;

/**
 * OFB 암호화 모드 클래스 입니다.
 */
public class OFB extends StreamBlockCipherMode {
    public OFB(@NonNull BlockCipher blockCipher) {
        super(EMode.ECB, blockCipher);
        this.blockSize = getBlockSizeByAlgorithm(this.algorithm);
    }

    @Override
    public Result encrypt(@NonNull byte[] data, @NonNull Param params) throws CryptoFailException {
        if (!(params instanceof ParamWithIV)) {
            throw new IllegalArgumentException("OFB 모드는 ParamsWithIV 객체를 요구합니다.");
        }
        ParamWithIV paramWithIV = (ParamWithIV) params;

        OFBBlockCipher cipher = new OFBBlockCipher(this.engine, this.blockSize);
        CipherParameters parameters = new ParametersWithIV(new KeyParameter(paramWithIV.getKey()), paramWithIV.getIv());
        cipher.init(true, parameters);

        return this.encryptBlock(data, cipher);
    }

    @Override
    public Result decrypt(@NonNull byte[] encryptedData, @NonNull Param params) throws CryptoFailException {
        if (!(params instanceof ParamWithIV)) {
            throw new IllegalArgumentException("OFB 모드는 ParamsWithIV 객체를 요구합니다.");
        }
        ParamWithIV paramWithIV = (ParamWithIV) params;

        OFBBlockCipher cipher = new OFBBlockCipher(this.engine, this.blockSize);
        CipherParameters parameters = new ParametersWithIV(new KeyParameter(paramWithIV.getKey()), paramWithIV.getIv());
        cipher.init(false, parameters);

        return this.decryptBlock(encryptedData, cipher);
    }
}

File: cryptography-chiper/src/main/kotlin/dev/retrotv/crypto/encryption/mode/OFB.kt
package dev.retrotv.crypto.encryption.mode

import dev.retrotv.crypto.encryption.block.BlockCipher
import dev.retrotv.crypto.encryption.param.Params
import dev.retrotv.crypto.encryption.param.ParamsWithIV
import dev.retrotv.crypto.encryption.result.Result
import dev.retrotv.crypto.enums.ECipher.*
import dev.retrotv.crypto.enums.EMode.*
import dev.retrotv.crypto.exception.CryptoFailException
import org.bouncycastle.crypto.engines.*
import org.bouncycastle.crypto.modes.OFBBlockCipher
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.ParametersWithIV

class OFB(blockCipher: BlockCipher) : CipherMode(ECB, blockCipher) {
    private val blockSize by lazy {
        when (this.algorithm) {
            AES, ARIA, LEA, SEED -> 128
            DES, TRIPLE_DES -> 64
            else -> throw IllegalArgumentException("사용할 수 없는 알고리즘 입니다.")
        }
    }

    @Throws(CryptoFailException::class)
    override fun encrypt(data: ByteArray, params: Params): Result {
        require (params is ParamsWithIV) { "OFB 모드는 ParamsWithIV 객체를 요구합니다." }

        // blockSize는 8 혹은 16만 입력 가능 (16 권장)
        val cipher = OFBBlockCipher(this.engine, this.blockSize)
            cipher.init(true, ParametersWithIV(KeyParameter(params.key), params.iv))

        val encryptedData = ByteArray(data.size)
            cipher.processBytes(data, 0, data.size, encryptedData, 0)

        return Result(encryptedData)
    }

    @Throws(CryptoFailException::class)
    override fun decrypt(encryptedData: ByteArray, params: Params): Result {
        require (params is ParamsWithIV) { "OFB 모드는 ParamsWithIV 객체를 요구합니다." }

        val cipher = OFBBlockCipher(this.engine, this.blockSize)
            cipher.init(false, ParametersWithIV(KeyParameter(params.key), params.iv))

        val originalData = ByteArray(encryptedData.size)
            cipher.processBytes(encryptedData, 0, encryptedData.size, originalData, 0)

        return Result(originalData)
    }
}